parl/layers/layer_wrappers.py

```diff
@@ -73,31 +55,150 @@ def check_or_replace_name(name, attr):
     return check_or_replace_name(new_name, attr)
 
 
+class LayerFunc(object):
+    def __init__(self, param_attr=False, bias_attr=False):
+        self.param_attr = param_attr
+        self.bias_attr = bias_attr
+
+    def sync_paras_to(self, target_layer, gpu_id):
+        """
+        Copy the paras from self to a target layer
+        """
+        ## isinstance can handle subclass
+        assert isinstance(target_layer, LayerFunc)
+        src_attrs = [self.param_attr, self.bias_attr]
+        target_attrs = [target_layer.param_attr, target_layer.bias_attr]
+
+        place = fluid.CPUPlace() if gpu_id < 0 \
+                else fluid.CUDAPlace(gpu_id)
+
+        for i, attrs in enumerate(zip(src_attrs, target_attrs)):
+            src_attr, target_attr = attrs
+            assert (src_attr and target_attr) \
+                or (not src_attr and not target_attr)
+            if not src_attr:
+                continue
+            src_var = fetch_var(src_attr.name)
+            target_var = fetch_var(target_attr.name, return_numpy=False)
+            target_var.set(src_var, place)
+
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        ## __new__ won't init the class, we need to do that ourselves
+        copied = cls.__new__(cls)
+        ## record in the memo that self has been copied to avoid recursive copying
+        memo[id(self)] = copied
+
+        ## first copy all content
+        for k, v in self.__dict__.items():
```

Comment:
use iteritems to prevent extra memory usage?

Reply:
Changed.